### PR TITLE
linked list implementation

### PR DESCRIPTION
**Pull request by SchruteFarmsCode**
own implementation of linkedlist

### CODE REVIEW

File: src/main/java/Main.java
Comment (pmat2) [inline src/main/java/Main.java:1]: Again do not change main

File: src/main/java/fundamentals/Node.java
Comment (pmat2) [inline src/main/java/fundamentals/Node.java:1]: Make this anonymous class

**Review by pmat2 — CHANGES_REQUESTED** [review]

File: src/main/java/fundamentals/Node.java
Comment (SchruteFarmsCode) [reply src/main/java/fundamentals/Node.java:1]: Removed class Node and implemented the Node interface as an anonymous class

File: src/main/java/Main.java
Comment (SchruteFarmsCode) [reply src/main/java/Main.java:1]: corrected

Comment (pmat2) [reply src/main/java/Main.java:1]: image

File: src/main/java/fundamentals/MyLinkedList.java
Comment (pmat2) [inline src/main/java/fundamentals/MyLinkedList.java:1]: move to `collections` package

File: src/main/java/fundamentals/MyLinkedList.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        while (current != null) {
+            if (current.getData().equals(data)) {
+                return current.getPrevious() != null ? current.getPrevious().getData() : null;
+            }
+            current = current.getNext();
+        }
+        return null;
+    }
+
+    private interface Node<T> {
```
Comment (pmat2) [inline src/main/java/fundamentals/MyLinkedList.java]: why interface?

File: src/main/java/fundamentals/MyLinkedList.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    private Node<T> first;
+    private Node<T> last;
+
+    public MyLinkedList() {
+        this.first = null;
+        this.last = null;
+    }
+
+    public void addFirst(final T data) {
+        Node<T> newNode = new Node<T>() {
```
Comment (pmat2) [inline src/main/java/fundamentals/MyLinkedList.java]: This is wrong on multiple layers. First abstraction layers are mixed. Second this tightly couples implementation of `Node` with list add method.

File: src/main/java/fundamentals/MyLinkedList.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            last = newNode;
+        } else {
+            newNode.setNext(first);
+            first.setPrevious(newNode);
+            first = newNode;
+        }
+    }
+
+    public void addLast(T data) {
+        Node<T> newNode = new Node<T>() {
```
Comment (pmat2) [inline src/main/java/fundamentals/MyLinkedList.java]: Same as above

**Review by pmat2 — CHANGES_REQUESTED** [review]
Needs structural rebuilding

File: src/main/java/Main.java
Comment (SchruteFarmsCode) [reply src/main/java/Main.java:1]: I don't change the main class - the problem appears again and I don't know what to do with it?

File: src/main/java/fundamentals/MyLinkedList.java
Comment (SchruteFarmsCode) [reply src/main/java/fundamentals/MyLinkedList.java:1]: moved to collections package

File: src/main/java/fundamentals/MyLinkedList.java (hunk 1 above)
Comment (SchruteFarmsCode) [reply src/main/java/fundamentals/MyLinkedList.java]: changed interface to static class

File: src/main/java/fundamentals/MyLinkedList.java (hunk 2 above)
Comment (SchruteFarmsCode) [reply src/main/java/fundamentals/MyLinkedList.java]: corrected

File: src/main/java/fundamentals/MyLinkedList.java (hunk 3 above)
Comment (SchruteFarmsCode) [reply src/main/java/fundamentals/MyLinkedList.java]: corrected

File: src/main/java/collections/MyLinkedList.java
Comment (pmat2) [inline src/main/java/collections/MyLinkedList.java:1]: Implement methods: `getFirst`, `getLast`, `get(index)`, `removeFirst`, `removeLast`, `remove(index)`, `clear`. You will have to add `index` to Node

Comment (SchruteFarmsCode) [reply src/main/java/collections/MyLinkedList.java:1]: Implement methods: getFirst(), getLast(), get(int index), removeFirst(), removeLast(), remove(int index), clear()

**Pull request merged**